hail/python/hail/nd/nd.py

```diff
@@ -314,8 +313,7 @@ def concatenate(nds, axis=0):
     """
     head_nd = nds[0]
     head_ndim = head_nd.ndim
-    for nd in nds:
-        assert(nd.ndim == head_ndim)
+    hl.case().when(hl.all(lambda a: a.ndim == head_ndim, nds), True).or_error("Mismatched ndim")
```

Comment:
Also, unfortunately hl.all does not work with tuples, which I think should be classified as bug @tpoterba